comms/dht/src/dedup/dedup_cache.rs

```diff
@@ -0,0 +1,184 @@
+
+const LOG_TARGET: &str = "comms::dht::dedup_cache";
+
+//#[derive(Clone, Debug, Queryable, Insertable, PartialEq)]
+#[derive(Clone, Debug, Queryable, Identifiable)]
+#[table_name = "dedup_cache"]
+struct DedupCacheSql {
+    pub id: i32,
+    pub body_hash: String,
+    pub number_of_hits: i32,
+    pub stored_at: NaiveDateTime,
+    pub last_hit_at: NaiveDateTime,
+}
+
+#[derive(Clone, Debug, Insertable, Default, AsChangeset)]
+#[table_name = "dedup_cache"]
+pub struct UpdateDedupCacheSql {
+    pub body_hash: Option<String>,
+    pub number_of_hits: Option<i32>,
+    pub last_hit_at: Option<NaiveDateTime>,
+}
+
+#[derive(Clone)]
+pub struct DedupCacheDatabase {
+    connection: DbConnection,
+    capacity: usize,
+}
+
+impl DedupCacheDatabase {
+    pub fn new(connection: DbConnection, capacity: usize) -> Self {
+        let capacity = max(capacity, 100);
+        debug!(
+            target: LOG_TARGET,
+            "Message dedup cache capacity initialized at {}", capacity,
+        );
+        Self { connection, capacity }
+    }
+
+    /// Inserts and returns Ok(true) if the item already existed and Ok(false) if it didn't
+    pub async fn insert_body_hash_if_unique(&self, body_hash: Vec<u8>) -> Result<bool, StorageError> {
+        let body_hash_string = hex::to_hex(&body_hash.as_bytes());
+        match self.insert_body_hash(body_hash_string.clone()).await {
+            Ok(val) => {
+                if val == 0 {
+                    warn!(
+                        target: LOG_TARGET,
+                        "Unable to insert new entry into message dedup cache"
+                    );
+                }
+                Ok(false)
+            },
+            Err(e) => match e {
+                StorageError::UniqueViolation(_) => match self.update_number_of_hits(body_hash_string).await {
+                    Ok(_) => Ok(true),
+                    Err(e) => Err(e),
+                },
+                _ => Err(e),
+            },
+        }
+    }
+
+    /// Trims the dedup cache to the configured limit by removing the oldest entries
+    pub async fn truncate(&self) -> Result<usize, StorageError> {
+        let capacity = self.capacity;
+        self.connection
+            .with_connection_async(move |conn| {
+                let mut num_removed = 0;
+                let msg_count = dedup_cache::table
+                    .select(dsl::count(dedup_cache::id))
+                    .first::<i64>(conn)? as usize;
+                // Hysteresis added to minimize database impact
+                if msg_count > capacity {
+                    let remove_count = msg_count - capacity;
+                    let message_ids: Vec<i32> = dedup_cache::table
+                        .select(dedup_cache::id)
+                        .order_by(dedup_cache::stored_at.asc())
+                        .limit(remove_count as i64)
+                        .get_results(conn)?;
+                    num_removed = diesel::delete(dedup_cache::table)
```

Comment:
You should combine the above select into this delete so that it is atomic (ideal), or otherwise wrap both in a transaction and change the select to lock the table for deletion (less ideal)

Reply:
Everything within `with_connection_async` is protected with an async Mutex. Can't remember now why I didn't implement SQL transactions (may have been time pressure). I think (but you'd need to check) that the DeleteStatement doesnt support limit, which causes an issue - otherwise, yeah they should ideally be in one query.